modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,529 @@
+    public static final int DFLT_FLUSH_SIZE = (int)(8 * U.MB);
+
+    /** Directory to store performance statistics files. Placed under Ignite work directory. */
+    public static final String PERFORMANCE_STAT_DIR = "performanceStatistics";
+
+    /** Factory to provide I/O interface. */
+    private final FileIOFactory fileIoFactory = new RandomAccessFileIOFactory();
+
+    /** Performance statistics enabled flag. */
+    private volatile boolean enabled;
+
+    /** Performance statistics file writer worker. */
+    @Nullable private volatile FileWriter fileWriter;
+
+    /** Performance statistics file I/O. */
+    private volatile FileIO fileIo;
+
+    /** File write buffer. */
+    @Nullable private volatile SegmentedRingByteBuffer ringByteBuffer;
+
+    /** Count of written to buffer bytes. */
+    private final AtomicInteger writtenToBuffer = new AtomicInteger();
+
+    /** {@code True} if the small buffer warning message logged. */
+    private final AtomicBoolean smallBufLogged = new AtomicBoolean();
+
+    /** {@code True} if worker stopped due to maximum file size reached. */
+    private final AtomicBoolean stopByMaxSize = new AtomicBoolean();
+
+    /** Hashcodes of cached strings. */
+    private final ConcurrentSkipListSet<Integer> cachedStrings = new ConcurrentSkipListSet<>();
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** @param ctx Kernal context. */
+    public FilePerformanceStatisticsWriter(GridKernalContext ctx) {
+        log = ctx.log(getClass());
+
+        this.ctx = ctx;
+    }
+
+    /** Starts collecting performance statistics. */
+    public void start() {
+        synchronized (this) {
+            if (enabled)
+                return;
+
+            enabled = true;
+
+            try {
+                File file = statisticsFile(ctx);
+
+                U.delete(file);
+
+                fileIo = fileIoFactory.create(file);
+
+                ringByteBuffer = new SegmentedRingByteBuffer(DFLT_BUFFER_SIZE, DFLT_FILE_MAX_SIZE,
+                    SegmentedRingByteBuffer.BufferMode.DIRECT);
+
+                ringByteBuffer.init(0);
+
+                fileWriter = new FileWriter(ctx, log);
+
+                new IgniteThread(fileWriter).start();
+
+                log.info("Performance statistics writer started [file=" + file.getAbsolutePath() + ']');
+            }
+            catch (IOException | IgniteCheckedException e) {
+                log.error("Failed to start performance statistics writer.", e);
+
+                stopStatistics();
+
+                throw new IgniteException("Failed to start performance statistics writer.", e);
+            }
+        }
+    }
+
+    /** Stops collecting performance statistics. */
+    public void stop() {
+        synchronized (this) {
+            if (!enabled)
+                return;
+
+            enabled = false;
+
+            FileWriter fileWriter = this.fileWriter;
```

Comment:
I can't see why we should cache this variable here?

Reply:
For example, if `fileIo = fileIoFactory.create(file);` throws an exception and the writer will not be created